lib/autokey/macro.py

```diff
@@ -30,6 +31,21 @@ def on_triggered(self):
     from gi.repository import Gtk
 
 
+def extract_tag(s):
+    if not isinstance(s, str):
+        raise TypeError
+    extracted = [p.split('>')[0] for p in s.split('<') if '>' in p]
+    if len(extracted) == 0:
+        return s
+    else:
+        return ''.join(extracted)
+
+def split_key_val(s):
+    # Split as if a shell argument.
+    # Splits at spaces, but preserves spaces within quotes.
+    pairs = shlex.split(s)
```

Comment:
Using shlex is an interesting idea.
When I started to work on this, I threw the biggest hammer on the problem I knew by the time, by building a full-fledged RE-based tokenizer and parser. My goal was to have a fully bug- and corner-case free parser. Unfortunately, there was no real end result. Just for reference, that’s the diff of my incomplete and stale attempt: https://github.com/autokey/autokey/compare/develop...luziferius:improve_phrase_macro_parsing

So let’s see how your much shorter solution copes with various corner cases.

Reply:
Can you tell me what sorts of corner cases you were referring to? Then I can add tests, at least.

Reply:
Sorry, I hadn't realised you had already put so much work in. I just came across needing spaces in args, and assumed there'd be a library to parse strings for arguments.

Reply:
Yours looks like it will probably be more sophisticated. What stage is it at? May be worth replacing my quick change with yours when it is done. Mine doesn't handle nested macros, and possibly never can.

Reply:
> Sorry, I hadn't realised you had already put so much work in. I just came across needing spaces in args, and assumed there'd be a library to parse strings for arguments.

No Problem! That branch lives inside my personal fork (see the diff link) and I consider it stale at the moment. It is incomplete. There’s no way I’d assume you look around my stuff to look for bits and pieces. So don’t worry about it.


> Can you tell me what sorts of corner cases you were referring to? Then I can add tests, at least.

Let’s see if I can still pull some from under my hat. I had mostly nesting broken macro tokens and similar stuff on my mind.


- `<script name="<script name="foo", args="None">, args="None">` This nests two valid macro tokens, if you consider `<script name=foo, args=None>` to be a valid script name (which it is!). Which one to execute is to decide…
- `<script <script name="myscript" args="valid_token">>`
- `<script name=<file <date format="%Y \"> some text <script (broken) <script name="myscript" args="valid_token">> args="<script name=`

Reply:
Ofc I wouldn't look at your bits, but neovim for example requires PRs discussing what you plan to do, to prevent duplication of work/unwanted work. At the least, we could set up some sort of communication system for what we are working on.

Reply:
Ah, they look quite involved. The second one currently isn't valid because the outer macro requires 'args', still.

TBH my method still doesn't handle nested macros anyway, let alone broken ones. I can look at implementing it, but it's lower priority unless you make it a requirement for merge.

Reply:
> Ah, they look quite involved. The second one currently isn't valid because the outer macro requires 'args', still.

In the second point, note that the outer `<script>`-thing has no `name` parameter. Here, the inner token is valid.

> TBH my method still doesn't handle nested macros anyway, let alone broken ones. I can look at implementing it, but it's lower priority unless you make it a requirement for merge.

That won’t be a requirement, so it’s fine as it is.